linera-execution/solidity/Linera.sol

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "./LineraTypes.sol";
+
+// This library provides Linera functionalities to EVM contracts
+// It should not be modified.
+
+// Precompile keys:
+// (0,0): chain_id
+// (0,1): application_creator_chain_id
+// (0,2): chain_ownership
+// (0,3): read data blob
+// (0,4): assert data blob exists
+// (1,0): try_call_application
+// (1,1): validation round
+// (1,2): send_message
+// (1,3): message_id
+// (1,4): message_is_bouncing
+// (2,0): try_query_application
```

Comment:
What's the reasoning for choosing this particular encoding? each element of the tuple is `uint8` so holds numbers up to 255. Wouldn't it be easier to simply keep the indexes monotonically increase?

Reply:
It has to match the bcs encoding of the type `PrecompileTag`.

Reply:
This organization of the `PrecompileTag` is the one before this PR.

The problem is that some runtime operations are available in both `Contract` and `Service`, some are available only in `Contract` and some only in `Service`.

With the proposed reorganization, things become much clearer and the code is better organized:
```rust
#[derive(Debug, Serialize, Deserialize)]
enum PrecompileTag {
    Base(BasePrecompileTag),
    Contract(ContractPrecompileTag),
    Service(ServicePrecompileTag),
}
```

But since we do not want to have boilerplate code, this implies having 2 bytes instead of just 1.

Reply:
I see. I still think it'd be good to have some better safety precautions here.

Maybe for the time being add a longer comment here that these indices have to match the serde encoding of Precompile tag. Also, maybe a comment on `PrecompileTag` with a warning that things should not be reorganised b/c they can break the integration.